### ISSUE TITLE
stream.mp3 is glitchy on a Pi Zero

### ISSUE THREAD

**Issue by beniamino38** [ejurgensen/forked-daapd#804]
I'm running forked-daapd on a Pi Zero. When I listen to the music through stream.mp3, the stream is unlistenably glitchy, even on a fast local network where airplay works seamlessly. The  issue might be that the Pi Zero is not fast enough to transcode the stream in real time. However, `top` shows the machine is 20% idle during streaming (with forked daapd running at a `nice` value of -20). I wonder if it might be possible to reduce CPU usage by tweaking the transcoding settings (along the lines of #503). Alternatively, I wonder if this could be due to how much priority is given to streaming within forked-daapd itself.

**Comment by whatdoineed2do** [comment]
Are you browsing your library whilst you are streaming the mp3?

On your PI Zero, trying looking at `top -H -p $(pidof forked-daapd)` and at the row with `httpd` in far right column - this is the thread running the HTTP daemon which is also where the streamed mp3 gets served.  Is the CPU% value high for this thread?  If so, it may be a similar issue to #793 

Regarding the transcode, the server transcodes the audio to 128kbps for streaming - there's a PR #789 that lets you adjust this but you'll need to clone the PR and the code (`httpd_streaming.c:628`) to specify a lower bit rate

**Comment by beniamino38** [comment]
Thanks for your comments. 

Using `top` I can see that the httpd thread is using 55%CPU when running at 128kbps. Using your PR and adjusting the bitrate down to 64kbps, the httpd thread uses 45% CPU and that seems low enough to get rid of the glitches completely. Lower bitrates aren't ideal for music but they are fine for audiobooks). Would you consider putting additional bitrate options into your PR to accommodate this (e.g. 64kbps, 96kbps)?

**Comment by whatdoineed2do** [comment]
Sure, I can add additional bit_rates - its obviously still pending any form of review so it may not get accepted though.

Additionally, there's another PR #802  that splits off the httpd serving and mp3 streaming into 2 different threads with their own event loops - it would be interesting if using PR #802  are you able to leave your streaming at 128kbps etc.  The downside is that you need to get your mp3 stream from a different, but configurable, port (3680) rather than the server default of 3689.

**Comment by whatdoineed2do** [comment]
@beniamino38 - I found my old RPi Zero and did some testing.

Running master vs code from PR #802 (separate httpd event loop for streaming) + #789 (select-able mp3 streaming bitrate) with 320kbps bitrate for the mp3 stream  Streaming was to my laptop using ffmpeg as the client (wifi) and during playback browsing the web ui.

* files (320kbps source files) local to RPi 0
`master`  and patched code both streamed with no glitches.

* files (320kbps source files) from a wired RPi 3 NFS mount
master suffered audiable glitches very more quickly but the patch wasnt totally immune (sending out a 320kbps stream) but very mcuh less frequent.  Dropping the streaming bit rate to 192kbps was flawless for streaming playback except when it appears the RPI was struggling to read the data from the NFS server

```
[2019-09-10 22:29:40] [  LOG]     scan: Bulk library scan completed in 30 sec (with file scan disabled)
[2019-09-10 22:29:40] [  LOG]      lib: Library init scan completed in 30 sec (0 changes)
[2019-09-10 22:30:32] [  LOG]   laudio: SNDRV_PCM_IOCTL_SYNC_PTR failed (-32)[2019-09-10 22:30:53] [  LOG]   player: Output delay detected (behind=2371, max=150), resetting all outputs
[2019-09-10 22:31:03] [  LOG]   laudio: SNDRV_PCM_IOCTL_SYNC_PTR failed (-32)[2019-09-10 22:31:04] [  LOG]   player: Output delay detected (behind=634, max=150), resetting all outputs
[2019-09-10 22:31:19] [  LOG]   player: Source is not providing sufficient data, temporarily suspending playback (deficit=295084/264600 bytes)
[2019-09-10 22:32:26] [  LOG]   laudio: SNDRV_PCM_IOCTL_SYNC_PTR failed (-32)[2019-09-10 22:32:36] [  LOG]   player: Output delay detected (behind=976, max=150), resetting all outputs
[2019-09-10 22:32:56] [  LOG]   laudio: SNDRV_PCM_IOCTL_SYNC_PTR failed (-32)[2019-09-10 22:33:03] [  LOG]   player: Output delay detected (behind=736, max=150), resetting all outputs
[2019-09-10 22:33:08] [  LOG]   player: Source is not providing sufficient data, temporarily suspending playback (deficit=264708/264600 bytes)
[2019-09-10 22:33:14] [  LOG]   player: Output delay detected (behind=215, max=150), resetting all outputs
[2019-09-10 22:33:16] [  LOG]   player: Output delay detected (behind=210, max=150), resetting all outputs
[2019-09-10 22:33:19] [  LOG]   player: Output delay detected (behind=264, max=150), resetting all outputs
[2019-09-10 22:33:21] [  LOG]   player: Output delay detected (behind=162, max=150), resetting all outputs
[2019-09-10 22:33:23] [  LOG]   player: Output delay detected (behind=232, max=150), resetting all outputs
[2019-09-10 22:33:28] [  LOG]   player: Source is not providing sufficient data, temporarily suspending playback (deficit=292068/264600 bytes)
```
.

Maybe you can try the code from PR #802 and see if that solves this issue but it looks like its a problem (at least on my RPI 0) is related to the wifi bandwidth on the device getting data in.

The `player` and `input` threads at this time were still in the 6-10% CPU util range whilst the thread serving the mp3 was 47-52%

**Comment by beniamino38** [comment]
Thanks for your insights! I think your assessment is right. On my Pi Zero, your `ffmpeg` test transcodes a typical file at about 2.5x the play speed. Which isn't as close as I feared, but may be close enough to cause problems if other threads are running.

I had a go at using libshine instead but it seems to require fixed-size frames, and I didn't see a way to achieve that.

I went back to libmp3lame and played with the parameters a bit, and found that `s->codec->compression_level = 9;` reduces the CPU usage of `forked-daapd` during streaming from ~59% to about ~42%, and that is enough to make streaming reliable for me.
